tezos/messages/src/p2p/binary_message.rs

```diff
@@ -26,208 +25,90 @@ pub const CONTENT_LENGTH_FIELD_BYTES: usize = 2;
-            }
-        };
-    }
+/// Trait for reading a binary message.
+pub trait BinaryRead: Sized {
+    /// Create new struct from bytes.
+    fn from_bytes<B: AsRef<[u8]>>(buf: B) -> Result<Self, BinaryReaderError>;
 }
 
-/// Trait for binary encoding to implement.
-///
-/// Binary messages could be written by a [`MessageWriter`](super::stream::MessageWriter).
-/// To read binary encoding use  [`MessageReader`](super::stream::MessageReader).
-pub trait BinaryMessage: Sized {
+/// Trait for writing a binary message.
+pub trait BinaryWrite {
     /// Produce bytes from the struct.
     fn as_bytes(&self) -> Result<Vec<u8>, BinaryWriterError>;
-
-    /// Create new struct from bytes.
-    fn from_bytes<B: AsRef<[u8]>>(buf: B) -> Result<Self, BinaryReaderError>;
 }
 
-impl<T> BinaryMessage for T
+/// Message that can be both encoded and decoded into binary format.
+pub trait BinaryMessage: BinaryRead + BinaryWrite {}
+
+impl<T: BinaryRead + BinaryWrite> BinaryMessage for T {}
+
+impl<T> BinaryWrite for T
 where
-    T: HasEncoding + cache::CachedData + DeserializeOwned + Serialize + Sized,
+    T: tezos_encoding::encoding::HasEncoding + Serialize,
 {
     #[inline]
     fn as_bytes(&self) -> Result<Vec<u8>, BinaryWriterError> {
-        // check cache at first
-        if let Some(cache) = self.cache_reader() {
-            if let Some(data) = cache.get() {
-                return Ok(data);
-            }
-        }
-
-        // if cache not configured or empty, resolve by encoding
         binary_writer::write(self, &Self::encoding())
     }
+}
 
+impl<T> BinaryRead for T
+where
+    T: tezos_encoding::nom::NomReader + Sized,
+{
     #[inline]
-    fn from_bytes<B: AsRef<[u8]>>(bytes: B) -> Result<Self, BinaryReaderError> {
-        let bytes = bytes.as_ref();
-
-        let value = BinaryReader::new().read(bytes, &Self::encoding())?;
-        let mut myself: Self = deserialize_from_value(&value)?;
-        if let Some(cache_writer) = myself.cache_writer() {
-            cache_writer.put(bytes);
-        }
-        Ok(myself)
+    fn from_bytes<B: AsRef<[u8]>>(buf: B) -> Result<Self, BinaryReaderError> {
+        all_consuming_complete_input(T::nom_read, buf.as_ref())
     }
 }
 
+/// This trait is able to predict the exact size of the message from the first bytes of the message.
+pub trait SizeFromChunk {
+    /// Returns the size of the message.
+    fn size_from_chunk(bytes: impl AsRef<[u8]>) -> Result<usize, BinaryReaderError>;
+}
+
+/// Applies nom parser `parser` to the input, assuming that input is complete.
+pub fn complete_input<'a, T>(
+    parser: impl FnMut(NomInput<'a>) -> NomResult<'a, T>,
+    input: NomInput<'a>,
+) -> Result<T, BinaryReaderError> {
+    // - `complete` combinator assumes that underlying parsing has complete input,
+    //   converting ``Incomplete` into `Error`.
+    // - `finish` flattens `nom::Err` into underlying error for easier processing.
+    let (_, output) = complete(parser)(input)
```

Comment:
@akoptelov 
I think this could be written like this:
```
complete(parser)(input)
    .finish()
    .map(|(_, output)| output)
    .map_err(|error| map_nom_error(input, error))
```

Reply:
Done. Also the next fn is changed accordingly.